components/support/jwcrypto/src/direct.rs

```diff
@@ -0,0 +1,227 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Support for "Direct Encryption with a Shared Symmetric Key"
+//! See https://tools.ietf.org/html/rfc7518#section-4.5 for all the details.
+
+use crate::{
+    aes,
+    error::{JwCryptoError, Result},
+    Algorithm, CompactJwe, EncryptionAlgorithm, JweHeader, Jwk, JwkKeyParameters,
+};
+use rc_crypto::rand;
+
+impl Jwk {
+    /// Create a new random key suitable for `Direct` symmetric encryption.
+    /// Consumers can store this for later use, probably via serde serialization.
+    pub fn new_direct_key(kid: Option<String>) -> Result<Self> {
+        // We only support AES256 which has a 32byte key.
+        let mut bytes: Vec<u8> = vec![0; 32];
+        rand::fill(&mut bytes)?;
+        Ok(Jwk {
+            kid,
+            key_parameters: JwkKeyParameters::Direct {
+                k: base64::encode_config(&bytes, base64::URL_SAFE_NO_PAD),
+            },
+        })
+    }
+
+    // Create a new Jwk given the raw bytes of a `Direct` key. We generally
+    // prefer consumers to use serde with the entire key and the only use-case
+    // is currently tests, so let's limit this to tests for now...
+    #[cfg(test)]
+    pub fn new_direct_from_bytes(kid: Option<String>, key: &[u8]) -> Self {
+        Jwk {
+            kid,
+            key_parameters: JwkKeyParameters::Direct {
+                k: base64::encode_config(&key, base64::URL_SAFE_NO_PAD),
+            },
+        }
+    }
+}
+
+pub(crate) fn encrypt_to_jwe(
+    data: &[u8],
+    enc: EncryptionAlgorithm,
+    jwk: &Jwk,
+) -> Result<CompactJwe> {
+    // It's slightly unfortunate we need to supply a struct with ECDH specific
+    // values all specified as None, but doesn't seem likely to ever actually hurt.
+    let protected_header = JweHeader {
+        kid: jwk.kid.clone(),
+        alg: Algorithm::Direct,
+        enc,
+        epk: None,
+        apu: None,
+        apv: None,
+    };
+    let secret = match &jwk.key_parameters {
+        JwkKeyParameters::Direct { k } => base64::decode_config(&k, base64::URL_SAFE_NO_PAD)?,
+        _ => return Err(JwCryptoError::IllegalState("Not a Direct key")),
+    };
+    aes::aes_gcm_encrypt(data, protected_header, &secret)
+}
+
+pub(crate) fn decrypt_jwe(jwe: &CompactJwe, jwk: Jwk) -> Result<String> {
+    let secret = match jwk.key_parameters {
+        JwkKeyParameters::Direct { k } => base64::decode_config(&k, base64::URL_SAFE_NO_PAD)?,
+        _ => return Err(JwCryptoError::IllegalState("Not a Direct key")),
+    };
```

Comment:
This is conceptually the layer that knows what to do with the `jwe.encrypted_key()` field.